src/runtime/task/task_queue.cpp

```diff
@@ -54,44 +61,46 @@ void task_queue::enqueue_internal(task *task)
 {
     auto &sp = task->spec();
     auto throttle_mode = sp.rpc_request_throttling_mode;
-    if (throttle_mode != TM_NONE) {
+
+    if (throttle_mode != TM_NONE || FLAGS_global_max_task_queue_length > 0) {
         int ac_value = 0;
         if (_spec->enable_virtual_queue_throttling) {
             ac_value = _virtual_queue_length;
         } else {
             ac_value = count();
         }
 
-        if (throttle_mode == TM_DELAY) {
-            int delay_ms =
-                sp.rpc_request_delayer.delay(ac_value, _spec->queue_length_throttling_threshold);
-            if (delay_ms > 0) {
-                auto rtask = static_cast<rpc_request_task *>(task);
-                if (rtask->get_request()->io_session->delay_recv(delay_ms)) {
-                    dwarn("too many pending tasks (%d), delay traffic from %s for %d milliseconds",
-                          ac_value,
-                          rtask->get_request()->header->from_address.to_string(),
-                          delay_ms);
-                }
-            }
-        } else {
-            dbg_dassert(TM_REJECT == throttle_mode, "unknow mode %d", (int)throttle_mode);
-
-            if (ac_value > _spec->queue_length_throttling_threshold) {
+        if (throttle_mode == TM_REJECT || FLAGS_global_max_task_queue_length > 0) {
+            if (ac_value >= std::min(_spec->queue_length_throttling_threshold,
+                                     FLAGS_global_max_task_queue_length)) {
                 auto rtask = static_cast<rpc_request_task *>(task);
                 auto resp = rtask->get_request()->create_response();
+                sp.on_rpc_task_reject_enqueue.execute(task);
                 task::get_current_rpc()->reply(resp, ERR_BUSY);
 
-                dwarn("too many pending tasks (%d), reject message from %s with trace_id = "
+                dinfo("too many pending tasks (%d), reject message from %s with trace_id = "
```

Comment:
dinfo_f

Reply:
What's the difference between `dinfo` and `dinfo_f`?

Reply:
use placeholder {} for int or string, but %d,%s,%....and so on